GPII-1682: Update "static" handler to support multiple content directories...

the-t-in-rtf opened the pull request:
See [GPII-1682](https://issues.gpii.net/browse/GPII-1682) for details.

the-t-in-rtf commented:
As we are getting a lot of these, I incorporated the fix for [GPII-1281](https://issues.gpii.net/browse/GPII-1281) into this PR.  IMO, we should confirm there is no other business on the earliest of these and work on this PR.

the-t-in-rtf commented:
In our latest discussion, we agreed that the "built in" support for namespaces in combination with component definitions is only coincidentally working and not supported.  The list of register components needs to be ordered manually within `gpii.express` before wiring everything together.

the-t-in-rtf commented:
@amb26, I have added a few static functions and integration tests and ensured that middleware and router ordering is handled correctly regardless of component ordering.

amb26 commented on tests/js/lib/request.js:
```diff
@@ -0,0 +1,17 @@
+var kettle = require("kettle");
+kettle.loadTestingSupport();
+
+fluid.defaults("gpii.express.tests.request", {
```
Could you follow the project standard ordering and place this instead in "gpii.test.express.request"? Cheers

the-t-in-rtf replied:
I am assuming you meant `gpii.tests.express`, which you used below and which is consistent with what I see in `infusion`.

amb26 replied:
See example on line 8 - the convention we have established is for "x.test" to refer to testing infrastructure common to all tests, and "x.tests" for actual test fixtures

the-t-in-rtf replied:
I may need you to break down what you mean by infrastructure and fixtures here.  For example, would the general caseHolder for a package be `gpii.test.package.caseHolder` and the one specific to a test be `gpii.tests.package.specificThingUnderTest.caseHolder`?  Please clarify.

amb26 replied:
Yes, that's right. If the thing can actually be concretely executed as a test fixture, it goes in "tests". If it can't, it is "test". Naturally this isn't a completely hard and fast rule if there are multiple layers of derivation between the framework and the fixture - but it's generally relatively clear if something is infrastructural (e.g. "kettle.test.request") or an actual test (e.g. "kettle.tests.KETTLE34dataSource")

the-t-in-rtf replied:
The latest commit addresses this, generic reusable bits are now in the `gpii.test` namespace, while specific testEnvironments and caseHolders are in the `gpii.tests` namespace.

amb26 commented on tests/js/lib/test-helpers.js:
```diff
@@ -1,10 +1,18 @@
-var fluid = require("infusion");
+var fluid = fluid || require("infusion");
 var gpii  = fluid.registerNamespace("gpii");
 fluid.registerNamespace("gpii.express.tests.helpers");
```
gpii.test.express.helpers is more consistent for this file, I think

the-t-in-rtf replied:
Done, throughout.

amb26 commented on tests/js/router/router-nesting-caseholder.js:
```diff
@@ -0,0 +1,62 @@
+require("../includes");
+require("./fixtures");
+
+fluid.defaults("gpii.express.tests.router.nesting.caseHolder", {
```
This one should be gpii.tests.express.router etc.

the-t-in-rtf replied:
Fixed throughout.  This will require some explicit changes in every package that builds tests on top of gpii.express.   Once we merge this, I will push that through our remaining holdings.

the-t-in-rtf commented:
@amb26, I have just committed a major refactor based on our discussion yesterday.  The contract for both routers and middleware is clearer, simpler, and hopefully the documentation is closer to what you had in mind.  Please review.

amb26 commented on docs/router.md:
```diff
@@ -0,0 +1,204 @@
+# About Routers
+
+`gpii.express.router` is a subclass of [`gpii.express.middleware`](middleware.md) that wraps
+[the  Express `router` object](http://expressjs.com/en/4x/api.html#router), which is simply a container for a collection
```
Need to explain what a router actually does - "simply a container for a collection of middleware" is a bit blase : P That is, need to mention our recently clarified concept of a router which is something that "provides routing based purely on information present in the URL path"

the-t-in-rtf replied:
We need to talk about this again.  Routers *may* isolate groups of middleware by path (including `/` for "any path") or by method (including `use` for "any method"), but they don't have to do either.  Middleware may also be limited by method or path, so the comparison also falls apart their.

The only difference is that a router may contain other middleware.  Say for example that you have a piece of `counter` middleware, and a `responder` piece of middleware.  

```
components: {
  get: {
    type: "gpii.express.router",
    options: {
      method: "get",
      components: {
        counter: { type: "my.namespace.counterMiddleware", priority: "first" },
        worker: { type: "my.namespace.respondingMiddleware"}
      }
    }
  },
  post: {
    type: "gpii.express.router",
    options: {
      method: "post",
      components: {
        counter: { type: "my.namespace.counterMiddleware", priority: "first" },
        worker: { type: "my.namespace.posRespondingMiddleware"}
      }
    }
  },
  default: {
    type: "gpii.express.router",
    options: {
      components: {
        counter: { type: "my.namespace.counterMiddleware", priority: "first" },
        worker: { type: "my.namespace.defaultRespondingMiddleware"}
      }
    }
  }
}
```

Each instance of `my.namespace.counterMiddleware` is isolated from the other instances, and can safely count just the traffic heading to the next piece of middleware in the chain.

Anyway, I'm not saying this shouldn't be clearer, only that we still need to chat, obviously.

amb26 replied:
Yes, as we established elsewhere, a "router" can also perform switching based on path and method. It's rather a shame that we're limited to express' primitives here - for example, the contentAwareMiddleware is only capable of switching to a collection of "handlers" rather than acting as a full-scale router (I'm assuming its former name of contentAwareRouter was simply a misnomer). This is fine, but it might be cool to establish something similar to Kettle's idea of "per-handler middleware", thus closing up one more point of difference between them

the-t-in-rtf replied:
Review the new docs in `handler.md` for a few ideas on alternate approaches and a fuller description of why we do things this way.  Hopefully that will give you an idea of how possible it is to move to another model.

amb26 commented on docs/router.md:
```diff
@@ -0,0 +1,204 @@
+grade.
+
+Routers can be nested, and you can even nest an instance of a router within itself.  The `path` variable for a given
+router module is combined with the paths of its parents, and that ultimately determines which URLs a given router will
```
More specific here - it is combined by suffixing it to the paths of its parents

the-t-in-rtf replied:
I updated the wording, take a look.

amb26 commented on docs/router.md:
```diff
@@ -0,0 +1,204 @@
+
+| Option          | Type                    | Description |
+| --------------- | ----------------------- | ----------- |
+| `method`        | `{String}`              | This grade provides the ability to limit itself to only operate on requests that match a particular [HTTP method](https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html).  Support lowercased string values, such as `get`, `post`, `put`, or `delete`. |
```
Indicate clearly which of these options are optional and which are required (and if they have any, what their defaults are)

the-t-in-rtf replied:
Do you have an example of how someone else has chosen to represent defaults and / or the fact that the field is required?

amb26 replied:
Here are a couple of examples from the Kettle docs:

https://github.com/amb26/kettle/tree/KETTLE-32#structure-of-a-kettle-config
https://github.com/amb26/kettle/tree/KETTLE-32#configuration-options-accepted-by-kettledatasourceurl

These aren't quite consistent with each other, but we should try to cobble together a standard pattern that resembles them

the-t-in-rtf replied:
I plan to use a carriage return and "(optional)" under optional fields, and to put the default (if any) at the end of the description.

the-t-in-rtf replied:
Just to illustrate the style I use for the future:

| Option              | Type       | Description |
| ------------------- | ---------- | ----------- |
| `requiredNoDefaults` | `{Number}` | A number that you have to set.  An error is thrown if this is not set. |
| `requiredWithDefaults`  |  `{Number}` | A number blah blah.  Set to `0` by default. |
| `optionalAlpha` (optional) | `{String}` | A string blah blah blah. |
| `optionalBeta` (optional) | `{String}` | A string blah blah blah. |

This is:

1. Required fields, alphabetically.
2. Optional fields, alphabetically, with "(optional)" after their `Option` value.
3. Defaults at the end of the `Description` value.

amb26 commented on src/js/contentAwareMiddleware.js:
```diff
@@ -0,0 +1,85 @@
+    events: {
+        onRequest: null
+    },
+    dynamicComponents: {
```
This is a near-duplicate of the setup in the "requestAware" grade. A cleaner and generalisable pattern would just set up the contract that the signature of the creator is (req, res, options) where options is a free-form space of options which will be applied to the request component's creator - cf. Kettle's "requests": https://github.com/amb26/kettle/blob/KETTLE-32/lib/KettleRequest.js#L23

the-t-in-rtf replied:
That example leaves me with a large question:  how can I have one argument which represents the component's options and have other arguments (req, res) that represent material that is meant to be a member defined within the component?  In that example, the bullet is dodged because we are only dealing with individual top-level entries in a component definition.

The only way I can think of is to use some kind of ugly wrapper, or to collapse everything into `options`, which doesn't provide any more clarity around what is expected than the existing model.

amb26 replied:
We had a chat about this and I referenced the location where we need a workaround for non-mergable definitions for component member entries - https://github.com/amb26/kettle/blob/KETTLE-32/lib/KettleRequest.js#L112 - this is caused by https://issues.fluidproject.org/browse/FLUID-5668

the-t-in-rtf replied:
We should review, I have a "last mile" problem with the `gradeNames` in the `options` argument not being merged in time.  The approach is otherwise sound.

amb26 replied:
Seems a priori impossible - all gradeNames are evaluated before the point of construction of the component. Perhaps you are seeing something else - happy to investigate

the-t-in-rtf replied:
Yes, I would have assumed it would work until I had clear evidence to the contrary.  Wrapping options in an enclosing object didn't help.  Anyway, I have only to uncomment a few lines to step through it together in our next pairing session.

the-t-in-rtf replied:
I added a link to the bug ([FLUID-5893](https://issues.fluidproject.org/browse/FLUID-5893)) in a TODO comment.

amb26 commented on docs/router.md:
```diff
@@ -0,0 +1,204 @@
+If multiple routers are registered for the same path, the first one to receive the request "wins".  As with middleware,
+you can control the order in which routers are added to the "chain" using
+[namespaces and priorities](http://docs.fluidproject.org/infusion/development/Priorities.html).
+The `path` option is one of two means by which a router can indicate that its scope is limited.  A router may also
```
Should indicate what is the name (`method`) of the option used to control this. Also should emphasise a little more (if this is true) that these means are not quite on an equal footing - one of them is optional and the other is required.

the-t-in-rtf replied:
Actually, they are both required, but have defaults that indicate "any method" (`use`) or "any path" (`/`).  I will try to clarify to address this and the above "provides routing based purely on information present in the URL path" comment, but we should talk first.

amb26 commented on src/js/errorMiddleware.js:
```diff
@@ -0,0 +1,63 @@
+ */
+gpii.express.middleware.error.sendError = function (that, error, request, response, next) {
+    if (response.headersSent) {
+        return next(error);
```
This is a little odd - firstly, who reads this return value - secondly, why don't we communicate an error specific to the situation rather than punting the original error - thirdly - if the headers have been sent, isn't it likely that any further attempt to communicate the error via the request will fail? This looks like a "log a high priority warning or terminate" situation to me. A few stray blank lines below

the-t-in-rtf replied:
I cleaned up this cycle.  There may still be downstream middleware (logging, etc.) that can still meaningfully do its work in such a case, so the error is still passed on.  I added a fluid.fail afterward, so that the error is clearly flagged.

amb26 commented on src/js/express.js:
```diff
@@ -1,44 +1,67 @@
+ * As maps do not preserve order, an array is always returned.  When used with a map, the keys will be discarded.
+ *
+ */
+gpii.express.orderByPriority = function (mapOrArray) {
```
This function is now covered by the framework utility fluid.parsePriorityRecords - https://github.com/fluid-project/infusion/blob/master/src/framework/core/js/Fluid.js#L1270 - or if it isn't, we should ensure that it is

the-t-in-rtf replied:
I tried this, but needed a newer version of Infusion to pick this up.  I updated to `2.0.0-dev.20160309T145652Z.9f515ec`, but got very odd errors.  When my debugging breakpoints in the infusion dependency weren't being hit, I figured out that the infusion from `kettle` was likely being used.  It seems to me like I need a newer matching set of `kettle` and `node-jqunit` to go with whatever `infusion` I use.  I looked briefly at the latest version of your KETTLE-32 branch, there is a different implementation of the priority ordering stack in the version used there.

Please help me identify a "matching set" for these three dependencies.

amb26 replied:
We looked into this and the apparent version mismatch was just a phenomenon of the IDE being confused by the module loading pattern triggered by Infusion's "auto-dedupe" feature not conforming to the node modules standard. The dedupe mechanism itself was working fine

the-t-in-rtf replied:
Ignore that, the problem turned out to be with the data itself, and we worked through it.  The only remaining problem is that priorities like `last` and `first` don't seem to work with handlers.  There is a relevant TODO in `contentAware-tests.js`.

amb26 commented on src/js/express.js:
```diff
@@ -64,98 +87,147 @@ gpii.express.registerComponentLineage = function (childComponent, expressCompone
+
+    // Order the combined data by priority so that we have predictable control over which is loaded when, then map it
+    // back to a simple ordered array.
+    return gpii.express.orderByPriority(nickNamesWithComponentPrioritiesAndNamespaces).map(gpii.express.extractNicks);
```
The function of "map" plus the custom function can be achieved by means of the framework utility fluid.getMembers - https://github.com/fluid-project/infusion/blob/master/src/framework/core/js/Fluid.js#L635

the-t-in-rtf replied:
Done.

amb26 commented on src/js/express.js:
```diff
@@ -64,98 +87,147 @@ gpii.express.registerComponentLineage = function (childComponent, expressCompone
-        component.router[component.options.method](component.options.path, component.route);
+        if (fluid.hasGrade(childComponent.options, "gpii.express.router")) {
+            // Recurse from here on down.
+            that.connectDirectDescendants(childComponent, path + "." + childNickname);
```
Don't need to necessarily fix this now, but should leave a TODO that this will fail in the near future when the framework supports periods in component names. This whole network of functions would be much safer if they operate with a "segs" array of path segments in place of "path" String

the-t-in-rtf replied:
I have added a TODO here.

amb26 commented on src/js/express.js:
```diff
@@ -64,98 +87,147 @@ gpii.express.registerComponentLineage = function (childComponent, expressCompone
+ *
+ * @param that {Object} - The `gpii.express` instance itself.
+ * @param component {Object} - The component we are working with at this level.  As we are treeing down through a hierarchy, this will change depending on the level we're at.
+ * @param path {String} - The path segment where we can fine the child components we need to wire in to ourselves.
```
fine -> find

the-t-in-rtf replied:
Fixed.

amb26 commented on src/js/headerMiddleware.js:
```diff
@@ -0,0 +1,69 @@
+        res.setHeader(headerOptions.fieldName, fieldValue);
+    });
+
+    if (err) {
```
Horrid, isn't it : P
You may like, for example. next.apply(null, fluid.makeArray(err)) but you may reject it as being "cutesy"

the-t-in-rtf replied:
Seems reasonable to me, IMO something like this stops short of "cutesy" as long as you add a meaningful comment in context (as I have done).

amb26 commented on src/js/requestAwareMiddleware.js:
```diff
@@ -0,0 +1,45 @@
+            target: "{that gpii.express.handler}.options.gradeNames"
+        }
+    ],
+    method:     "get",
```
stray spaces

the-t-in-rtf replied:
The entire line was left over from the previous implementation, and caused problems elsewhere.  I have removed it entirely, the grade will now correctly pick up the default `use` method from `gpii.express.middleware`.

amb26 commented on src/js/requestAwareMiddleware.js:
```diff
@@ -0,0 +1,45 @@
+        "onRequest": null
+    },
+    dynamicComponents: {
+        requestHandler: {
```
See comments on contentTypeAware for reforming this to be a single standard definition

the-t-in-rtf replied:
Done.

amb26 commented on src/js/router.js:
```diff
@@ -1,39 +1,51 @@
+            args:     ["{that}"]
+        },
+        middleware: {
+            funcName: "fluid.identity"
```
Not a great choice since fluid.identity dispatches only one argument. Perhaps fluid.notImplemented (i.e. writing nothing) would be closer to required intent - but I'm worried since fluid.identity here suggests that there is nothing that tests this default (or tests it in a way that doesn't check that "response" is dispatched too)

the-t-in-rtf replied:
It's not a default, it's sidestepping the requirement imposed by `gpii.express.middleware`, as I couldn't figure out how to tell it that the `router` member will in fact satisfy the contract for the `middleware` invoker.  Suggestions welcome.

the-t-in-rtf replied:
I ended up making the requirement for a middleware invoker or member a part of the contract enforced by the `getMiddlewareFunction` function in `gpii.express.middleware`, which will fail if it can't find a member or invoker.

amb26 commented on tests/js/errorMiddleware/errorMiddleware-tests.js:
```diff
@@ -0,0 +1,149 @@
+                            namespace: "simpleErrorPitcher",
+                            priority: "first",
+                            path: "/simple",
+                            components: {
```
This empty block can be omitted?

the-t-in-rtf replied:
Thanks, I obviously lost the ability to see the detritus leftover from the previous code.   It's gone now.

amb26 commented on tests/js/priority/priority-integration-tests.js:
```diff
@@ -0,0 +1,186 @@
+    }
+    else {
+        var headerValue = response.get("Exquisite-Corpse");
+        response.send((headerValue  || "") + string);
```
Stray space

the-t-in-rtf replied:
Fixed.

amb26 commented on tests/js/router/index.js:
```diff
@@ -0,0 +1,8 @@
+require("./router-wildcard-tests");
+
+// TODO:  This test causes spurious "assertion outside test context" in other tests if it's run at the same time.  Discuss with Antranig.
+//require("./router-static-failure-tests");
```
Shouldn't we try to run it somehow all the same?
Oddly, it seems that each of these test files are trying to construct the test environment synchronously on the spot rather than queuing it properly by using the fluid.test.runTests driver - http://docs.fluidproject.org/infusion/development/IoCTestingFramework.html#-modules-tests-and-sequence-

the-t-in-rtf replied:
None of the tests you have reviewed to date have used that, but I wish we had talked about it earlier, as that does seem much cleaner and lends itself to easier reuse.  My one question is how to structure it so that I can easy run a subset of tests (right now I can run the individual file that instantiates the test environment).

the-t-in-rtf replied:
Nevermind, I just realized that you meant that I should use `fluid.test.runTests` in the individual files.

the-t-in-rtf replied:
Done.

amb26 commented on docs/contentAwareMiddleware.md:
```diff
@@ -0,0 +1,62 @@
+if possible, and allows that to handle the original request.  If no handler can be found, `next(err)` is called and
+downstream [error handling middleware](middleware.md) is expected to handle things from there.
+
+* `reques`: The [request object](http://expressjs.com/en/api.html#req) provided by Express, which wraps node's [`http.incomingMessage`](https://nodejs.org/api/http.html#http_class_http_incomingmessage).
```
It doesn't wrap it, it *is* the node server's object - which is derived from the http.incomingMessage prototype

the-t-in-rtf replied:
Done.

amb26 commented on docs/contentAwareMiddleware.md:
```diff
@@ -0,0 +1,62 @@
+downstream [error handling middleware](middleware.md) is expected to handle things from there.
+
+* `reques`: The [request object](http://expressjs.com/en/api.html#req) provided by Express, which wraps node's [`http.incomingMessage`](https://nodejs.org/api/http.html#http_class_http_incomingmessage).
+* `response`: The [response object](http://expressjs.com/en/api.html#res) provided by Express, which wraps node's [`http.ServerResponse`](https://nodejs.org/api/http.html#http_class_http_serverresponse).
```
As previous

the-t-in-rtf replied:
I have moved the common wording to a standard location and updated it based on your feedback.

amb26 commented on docs/errorMiddleware.md:
```diff
@@ -0,0 +1,61 @@
+### `{that}.middleware(error, request, response, next)`
+
+* `error`: The raw error reported by the middleware itself.
+* `request`: The [request object](http://expressjs.com/en/api.html#req) provided by Express, which wraps node's [`http.incomingMessage`](https://nodejs.org/api/http.html#http_class_http_incomingmessage).
```
See above

the-t-in-rtf replied:
Done.

amb26 commented on docs/handler.md:
```diff
@@ -0,0 +1,63 @@
+
+| Option     | Type       | Description |
+| ---------- | ---------- | ----------- |
+| `request`  | `{Object}` | The [request object](http://expressjs.com/en/api.html#req) provided by Express, which wraps node's [`http.incomingMessage`](https://nodejs.org/api/http.html#http_class_http_incomingmessage). The `requestAware` middleware grade and its derivatives (such as the `contentAware` middleware) set this for you. |
```
See above - these bits of duplicate text should just link to a single source where they are defined once and for all

the-t-in-rtf replied:
Done.

amb26 commented on docs/headerMiddleware.md:
```diff
@@ -0,0 +1,171 @@
+
+### `{middleware}.middleware(req, res, next)`
+
+* `req`: The [request object](http://expressjs.com/en/api.html#req) provided by Express, which wraps node's [`http.incomingMessage`](https://nodejs.org/api/http.html#http_class_http_incomingmessage).
```
See above

the-t-in-rtf replied:
Done, throughout.

amb26 replied:
The text here still looks the same to me

the-t-in-rtf replied:
You are correct.  I missed it in my previous cull because the variable names were also different.  I have searched throughout for both variations, this was the only holdout, and is fixed.

amb26 commented on docs/middleware.md:
```diff
@@ -0,0 +1,173 @@
+
+| Option      | Type                    | Description |
+| ----------- | ----------------------- | ----------- |
+| `method`    | `{String}`              | This grade provides the ability to limit itself to only operate on requests that match a particular [HTTP method](https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html).  Support lowercased string values, such as `get`, `post`, `put`, or `delete`. |
```
As above, indicate which are optional, and if so, if there is a default value

the-t-in-rtf replied:
Done, throughout.

amb26 commented on docs/router.md:
````diff
@@ -0,0 +1,204 @@
+Let's say we have an instance of `gpii.express` that is configured as follows:
+
+```
+gpii.express({
````
I would prefer not to advertise this bad practice in the docs - since you know that people will just cut and paste! The argument supplied to gpii.express should be defined as a grade with gpii.express as a parent - and then the construction done as a separate act. This avoids mixing directives to the RES POTENTIA and the RES EXTENSA (explanation available on request)

the-t-in-rtf replied:
It's a fair point.  I have updated the examples.

the-t-in-rtf commented:
There were various comments about updating the docs with default values and an indication about which fields are required.  I have done that.

the-t-in-rtf commented:
From a chat in [#fluid-tech](https://botbot.me/freenode/fluid-tech/) yesterday:

> @amb26: the-t-in-rtf has been working with the utility kettle.test.pushInstrumentedErrors
> @amb26: And has the following implementation: https://github.com/the-t-in-rtf/gpii-express/blob/GPII-1682/tests/js/errorMiddleware/errorMiddleware-caseholder.js#L100
> @amb26: We found that his handler wasn't being called
> @amb26: This is because the primary purpose of this method is to allow testing of an "uncaught exception handler"
> @amb26: In his case, since he is in an express handler chain, express will always catch any propagating exceptions
> @amb26: What he was wanting to do was to explicitly check the action of fluid.fail 5:52
> @the-t-in-rtf: Actually, not exactly.
> @amb26: Here is the impl of pushInstrumentedErrors: https://github.com/amb26/kettle/blob/KETTLE-32/lib/test/KettleTestUtils.js#L87
> @the-t-in-rtf: unless by that you mean, confirm that fluid.fail is called when I wander down a broken road.
> @amb26: the-t-in-rtf - I did mean that
> @amb26: So, this function has two effects - firstly, to displace the jqUnit-scoped default handler for fluid.fail which immediately fails the current test
> @amb26: secondly, to register a global uncaught exception handler
> @amb26: What the-t-in-rtf wants isn't quite either of these
> @amb26: But instead to register his own handler as by this line: https://github.com/amb26/kettle/blob/KETTLE-32/lib/test/KettleTestUtils.js#L89
> @amb26: That is, to displace the automatic "test-failing" listener for fluid.fail by a custom test assertion 
> @amb26: So what's required here is to add his custom test assertion here in the role for what currently receives fluid.identity
> @the-t-in-rtf: OK, great, I see that now. the-t-in-rtf @ 5:56	
> @amb26: And then when the test is done, to pop it off the stack again with removeListener: https://github.com/amb26/kettle/blob/KETTLE-32/lib/test/KettleTestUtils.js#L95

I have updated the `errorMiddleware` tests based on this discussion.  We should check in, but as far as I can see, the work here is far enough along to safely push the recent changes through to other packages.  There will likely be another few rounds of pushing things through, but it'll reduce the number of changes that have to be made and tested at once.

the-t-in-rtf commented:
@amb26, I just reviewed all comments to make sure I had addressed your feedback to date.  As far as I can see only the [discussion around handlers](https://github.com/GPII/gpii-express/pull/20#discussion_r60205508) still needs to be wrapped up.

amb26 commented on docs/contentAwareMiddleware.md:
````diff
@@ -0,0 +1,62 @@
+}
+```
+
+Note that the nonsensical `contentType` value `default` in the above example is required.  Entries without a
````
Why provide a nonsensical one rather than an actual one?

the-t-in-rtf replied:
The implementation of `request.accepts` skips falsy values.  We could add code to change this, as long as we agree what should happen when it is omitted.  IMO, omitting the contentType should mean "accept any content".

the-t-in-rtf replied:
I reviewed [the code that handles `req.accepts`[(https://github.com/jshttp/accepts) to confirm, it matches the first entry if there are no `Accept` headers, which means that even though omitting `contentType` should "match anything", we still need a clear means of handling cases in which there are no `Accept` headers.  The previous approach was adapted and expanded in both the docs and the tests.   Hopefully it's clearer now.

amb26 commented on docs/errorMiddleware.md:
```diff
@@ -0,0 +1,61 @@
+* `error`: The raw error reported by the middleware itself.
+* `request {Object}` An Express Request object (see [the docs](request.md) for details.
+* `response {Object}` An Express Response object (see [the docs](response.md) for details.
+* `next`: The next piece of middleware in the error reporting chain. Unused in this implementation.
```
In what way is it unused? Can't the user call it? And if not, what effects would it have? If it can't be usefully called, it should be removed

the-t-in-rtf replied:
This was just wrong, we do actually use it.  I've updated the docs to remove the inaccurate information.

amb26 commented on docs/errorMiddleware.md:
```diff
@@ -0,0 +1,61 @@
+
+This invoker fulfills the standard contract for a `gpii.express.middleware` component, with the signature expected for
+an error handler.  Note that although our implementation does not make use of the `next` argument, we include it in our
+invoker signature to ensure that express correctly considers us as error handling middleware.
```
Given you've got a wrapper applied to this function, couldn't it be removed from the apparent signature which the user receives?

the-t-in-rtf replied:
It's actually used now, I removed the inaccurate wording.

amb26 commented on docs/handler.md:
```diff
@@ -0,0 +1,122 @@
+
+The approach used in this package is to dynamically create a "handler" component for each request.  Because a handler
+is only ever used for a single request, it can safely work with its own model and/or member data, including making use
+of  [change appliers and model listeners](http://docs.fluidproject.org/infusion/development/ChangeApplierAPI.html).
```
And, most importantly, events

the-t-in-rtf replied:
Good point, I updated the wording here.

amb26 commented on docs/headerMiddleware.md:
```diff
@@ -0,0 +1,175 @@
+                    }
+                }
+            }
+            // TODO:  You, the reader, should add at least one middleware component here that will respond to the user.
```
Or a handler?

the-t-in-rtf replied:
No.  Our handlers do not yet work like those in kettle.  They require an enclosing piece of middleware, one that has `gpii.express.handlerDispatcher` mixed into its `gradeNames`.

amb26 commented on docs/requestAwareMiddleware.md:
```diff
@@ -0,0 +1,45 @@
+# The "Request Aware" middleware.
+
+Some middleware (such as the "static" middleware included in this package) do not keep any kind of state variables that
```
Should be clearer - ALL middleware must not keep any state variables which can be exposed between requests

the-t-in-rtf replied:
Actually, that's not the case.  The `counter` middleware used in the tests is an example of middleware that needs to keep state variables between requests.  You might also have middleware that is coordinating state between users, such as in a multiplayer game or shared editing environment.

the-t-in-rtf replied:
We agreed to clarify this that no state variables should be saved in middleware, and that this grade adds an additional level of protection to ensure that this can happen while still being able to use x... y... and z from the framework.

amb26 commented on docs/router.md:
```diff
@@ -0,0 +1,216 @@
+[the  Express `router` object](http://expressjs.com/en/4x/api.html#router), which is simply a container for a collection
+of [middleware](middleware.md).
+
+Both Express and the Express `router` object have the concept of a "stack" of middleware.  A `gpii.express.router`
```
In light of our last week's conversation, all of this discussion should be rewritten to refer to a "tree" rather than a "stack" of middleware

the-t-in-rtf replied:
Within the context of each of these "containers", there is an actual `stack` variable, which is why I chose the name.

the-t-in-rtf replied:
I want to leave a reference to `stack` in there if possible.  I've added a block expanding on the previous explanation, which includes:

1. The "tree" wording
2. A link to the Wikipedia article we discussed.
3. The diagram from the Wikipedia article that illustrates the "pre-order" strategy used.
4. A sample component options block that matches the diagram.

Please review, hopefully this is not far from something we can both live with.

amb26 commented on src/js/contentAwareMiddleware.js:
```diff
@@ -0,0 +1,68 @@
+    var orderedHandlers = fluid.parsePriorityRecords(that.options.handlers, "Content aware entry");
+    // var orderedHandlers = gpii.express.orderByPriority(that.options.handlers);
+    var matchingHandler = fluid.find(orderedHandlers, function (value) {
+        return Boolean(value.contentType && request.accepts(value.contentType)) ? value : undefined;
```
Doesn't seem necessary to convert to boolean here

the-t-in-rtf replied:
Fixed.

amb26 commented on tests/js/priority/priority-getOrderedMemberNames-unit-tests.js:
```diff
@@ -0,0 +1,55 @@
+    jqUnit.assertDeepEq("The nicknames should be in the correct order now...", ["alpha", "beta", "charlie"], output);
+});
+
+jqUnit.test("A list of nicknames should remain in the same order if it has no namespaces or priorities...", function () {
```
This test is invalid. Order of iteration of object keys is undefined in versions of JS before ES6.

the-t-in-rtf replied:
The code under test no longer exists and was removed in favor of direct calls to `fluid.parsePriorityRecords`.

the-t-in-rtf commented:
I  have updated `infusion` to pick up the fix for [FLUID-5893](https://issues.fluidproject.org/browse/FLUID-5893), and removed the previous workaround.

cindyli commented on docs/router.md:
```diff
@@ -0,0 +1,300 @@
+                                a: {
+                                    type: "my.namespaced.middleware"
+                                },
+                                b: {
```
Is this component name `b` ambiguous with the same `b` component at line 48? Is this intentional? Why?

the-t-in-rtf replied:
Nope.  It was a mistake.  We fixed it during the call.

cindyli commented on README.md:
```diff
@@ -1,109 +1,62 @@
+As you can see, you are expected to have a `config.express` option that includes at least a `port` and `baseUrl`
+setting.  See the [documentation for the `gpii.express` grade](./docs/express.md) for a full list of configuration
+options.  This example configures a "static" router that is designed to serve up filesystem content (see
+["Static Router Module"](#static-router-module) for more details).
```
The link for "Static Router Module" is broken.

the-t-in-rtf replied:
Thanks very much for commenting, I appreciate the feedback.  This is fixed now.
